Fixing synthesizing keys for multiple keys pressed down on flutter web

erickzanardo opened the pull request:
## Description

The current implementation of the keyboard on the web engine had a logic to automatically trigger keyup events, when repeat events stoped occurring for a time.

That works great when only one key is pressed down, but when two keys are being pressed down together, the first key will stop to send repeat events and only the last key been pressed down will send repeat events.

That was causing a strange behaviour, as a keyup event would be triggered, even if the key was still pressed down.

I did some investigation, and due to some comments on the code, I discovered that this logic that auto triggers keyup when repeat event cease, was created to prevent the framework be on a invalid state due to a keyup event not happening because the browser don't trigger keyups when a key is used on a browser/system shortcut. So my solution was built to that same idea, and I add a small new logic, that only synthesise keyup events when a key can be used in a shortcut, which is when a modifier (control, alt, shift or meta) key is pressed together.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/58727

## Tests

- `test/keyboard_test.dart`

## Checklist

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

erickzanardo commented:
/cc: @mdebbar

erickzanardo commented:
> I couldn't find any tests for this file, I tried looking on the test folder but apparently there is no test for this class.

I just realised that I just failed to find the test file, but it exists, here it is: https://github.com/erickzanardo/engine/blob/bec9c89fabfecf4ef529ed90214740bfaa6aa52f/lib/web_ui/test/keyboard_test.dart

Will try to work on the test case for this change soon

mdebbar commented:
Apologies for the late response, I missed the email notification for the PR.

Thanks for working on this issue. Yes, the correct place to add tests for this is in `test/keyboard_test.dart`. You'll find plenty of examples there that you can learn and copy from.

This PR is causing one test to fail: `"do not synthesize keyup when we receive repeat events"` in `test/keyboard_test.dart`. Could you please take a look and see why it's failing? Let me know if you need help.

erickzanardo commented:
> Apologies for the late response, I missed the email notification for the PR.
> 
> Thanks for working on this issue. Yes, the correct place to add tests for this is in `test/keyboard_test.dart`. You'll find plenty of examples there that you can learn and copy from.
> 
> This PR is causing one test to fail: `"do not synthesize keyup when we receive repeat events"` in `test/keyboard_test.dart`. Could you please take a look and see why it's failing? Let me know if you need help.

Sure thing, I will be working on the tests for this PR someday this week, I will report back here if I succeed or if I have any issue, thanks!!

erickzanardo commented:
@mdebbar it took me some time to be able to get back to this sorry.

I have added the a new test and fixed the one breaking.

That test was breaking because since with this change, only keys affected by the meta modifiers are synthesised, triggering repeat events don't cancel that synthesizing.

From what I tested on the navigators, when a short is pressed, repeat events should not follow, so this should not be any problems, nevertheless I changed the test to trigger meta modified repeat events, in the case I missed some case in my tests on the browsers.

Please, let me know what do you think.

mdebbar commented on lib/web_ui/test/keyboard_test.dart:
```diff
@@ -453,6 +453,7 @@ void main() {
             key: 'i',
             code: 'KeyI',
             repeat: true,
+            isMetaPressed: true,
```
This test is trying to simulate the case where the meta keys have been released. I think we shouldn't use `isMetaPressed: true` because the meta key has been released already (see line 436 above).

mdebbar commented on lib/web_ui/test/keyboard_test.dart:
```diff
@@ -464,7 +465,7 @@ void main() {
             'key': 'i',
             'code': 'KeyI',
-            'metaState': 0x0,
+            'metaState': 0x08,
```
Same thing here. Meta keys have been released so this should remain `0x0`.

mdebbar commented on lib/web_ui/lib/src/engine/keyboard.dart:
```diff
@@ -91,9 +91,9 @@ class Keyboard {
-    if (!_isModifierKey(event)) {
+    // Don't synthesize a keyup event for modifier keys, or keys not affected by modifiers,
+    // because the browser always sends a keyup event for those.
+    if (!_isModifierKey(event) && _isAffectedByModifiers(event)) {
```
To fix the issue that's making tests fail, let's move the new condition inside. This way we guarantee that the syntehsizing timer is always canceled.

Does something like this work:
```dart
if (!_isModifierKey(event)) {
  _keydownTimers[timerKey]?.cancel();
  if (event.type == 'keydown' && _isAffectedByModifiers(event)) {
    _keydownTimers[timerKey] = Timer(_keydownCancelDuration, () {
      _keydownTimers.remove(timerKey);
      _synthesizeKeyup(event);
    });
  } else {
    _keydownTimers.remove(timerKey);
  }
}
```

erickzanardo replied:
I remembering trying that, but that didn't solved the issue (flutter/flutter#58727), but right now, I can't remember why,  I will give this another try, because looking again, this suggestion should work! Maybe I have confused things.

mdebbar replied:
@erickzanardo did you have a chance to try it out?

erickzanardo replied:
Hey, I haven't yet, sorry, I have been really overwhelmed at work. But I am planning on getting back to this PR on this saturday or sunday.

mdebbar replied:
No rush.

If the suggestion doesn't work, please ping so we can discuss it further.

mdebbar commented:
@erickzanardo your branch is based on an old commit in master. Could you please pull master and rebase your branch? Then push your branch.

Once that's done, we'll wait for all the tests to pass then I'll merge the PR.

Thanks again for the contribution!

erickzanardo commented:
> @erickzanardo your branch is based on an old commit in master. Could you please pull master and rebase your branch? Then push your branch.
> 
> Once that's done, we'll wait for all the tests to pass then I'll merge the PR.
> 
> Thanks again for the contribution!

Sure thing, will do that tonight!

erickzanardo commented:
@mdebbar Rebased with master! Let me know if that is ok.

Thanks

erickzanardo commented:
@mdebbar I notice that some tests will failed, which seems unrelated to the change on this PR? Let me know if a new rebase would be required

mdebbar commented:
You are right. Both are unrelated. There's no need to rebase at the moment.

- `luci-engine`: This represents the state of the tree. It'll become green as soon as the tree is green.
- `Mac iOS Engine`: It's likely a flaky failure, so I kicked off a re-run.

Once the re-run passes, and the `luci-engine` goes green, we can merge the PR.

erickzanardo commented:
> You are right. Both are unrelated. There's no need to rebase at the moment.
> 
> * `luci-engine`: This represents the state of the tree. It'll become green as soon as the tree is green.
> * `Mac iOS Engine`: It's likely a flaky failure, so I kicked off a re-run.
> 
> Once the re-run passes, and the `luci-engine` goes green, we can merge the PR.

Great, thanks for clarifying it!
